### DISCUSSION TITLE
Minify/optimize code by default - NEVERMIND, JUST SAW THIS IS A PRO FEATURE :)

### DISCUSSION THREAD

**Discussion by nickberens360** [cssbattle/forum#28] [Ideas]
It would be nice if the minification of the code was handled automatically upon submit. It gets really tedious to remove white space, the last semi-colon etc by hand when you're done with a challenge. I never do it because it just feels like a chore. Using short class names and p tags instead of divs all made sense but the base minification should be a baseline if that makes sense? Regardless it's a super fun app!

**Comment by gunnarbittersmann** [comment]
Editor add-ons are available with the [pro account](https://cssbattle.dev/pro/).

**Comment by Ariane-B** [comment]
I'd like to come back on this and I think it would be worthwhile to consider minifying code by default, regardless of pro status. Since the number of characters is a factor, without that, we're basically actively encouraging people to write bad/unclear code (no classes, single-character classes, everything on the same line, no comments, etc.), which kinda hurts the educational usefulness for everyone.

Other editor niceties I of course still think can be pro features without any issues, but specifically because the number of characters is a criterion, minification would be good.

That, OR we could alter the score calculation so that the number of characters isn't important, and what matters is the number of rules/properties/HTML tags. Bit more complex, but it may be advantageous.

**Reply by gunnarbittersmann** [reply]
> OR we could alter the score calculation so that the number of characters isn't important, and what matters is the number of rules/properties/HTML tags.

Keep in mind that you can solve any target with just one rule with one declaration: `body { background: … }`.

**Reply by Ariane-B** [reply]
Do you mean a background image? If you do that, won't you get a message saying images are forbidden?

**Reply by gunnarbittersmann** [reply]
No. Well, yes – in some kind. I didn’t mean _external_ background images. But there are _internal_ ones: gradients.

Here’s a [solution to #57](https://codepen.io/gunnarbittersmann/pen/mdGVbWv?editors=0100) using just `body { background: … }`. Removing whitespace, selecting `*` instead of `body`, using other length units and a radial-gradient optimization, I could bring this down to 697 charaters – out of the top 100.

This way you might even set pixel by pixel. This wouldn’t lead to a short solution in terms of characters, though – quite contrary, it would be huge. But it would be short in terms of numbers of rules and declarations.

**Reply by Ariane-B** [reply]
Oh. I hadn't considered that. Yeah number of rules may not be the smartest thing after all.

Or you'd need to add in exceptions for the background and at that point it's getting rather complex.

**Comment by Beowolve** [comment]
The purpose of CSSBattle is "not" to write good code but short code. I prefer the way it is now because there is a good chance that the used algorithm to minimize would mess with the code.
Nothing prevents you from using classes and good coding style but in that case you are just not going for short code anyway. I would say, its better to keep your spaces if you prefer good code over short code.
I personally have forgotten how to press the space bar and classes, whats that? 😄

**Reply by Ariane-B** [reply]
What do you mean, the minifier would mess with the code? I've never seen that happen, but maybe there's just something I'm not understanding?

As for preferring short code... You know, auto-minifying wouldn't keep you from writing pre-minified code if you wanted. I don't think the user experience of someone who enjoys forgetting the existence of the spacebar would change much, would it?

Anyway my main point is that regardless of whether CSS Battle's purpose is educative, its feature that goes "here's an example; let's see how close you can get to it with a comparison tool" can be a really good environment for juniors/people who originally did more back-end to learn CSS. 

But the number of characters thing tends to make people focus on the wrong details for improving their scores (not simpler styles but minifying everything into one line).

**Reply by Beowolve** [reply]
I had the situation where the default minifier from the PRO plugin changed the code so, that it didn't work anymore. So this can be an issue and shouldn't be a default setting in my opinion.

**Reply by Ariane-B** [reply]
That sounds like an issue with that specific minifier, not with minifiers in general. 

What you describe can possibly happen with advanced optimization features like rule simplification and stuff. But a simple one that only does the basics shouldn't even be capable to cause that. 

We have a whole app that gets automatically minified (in a simple way) in the build process, and never have styles been altered at all.

**Reply by Beowolve** [reply]
I believe you, but will it also work if the code that has invalid CSS/HTML code? To save the last chars and get a top score it is standard to not use end tags or dismiss closing quotation marks where possible for example. I expect many tools to fail on such code and I guess that is what happend back then when I still had PRO...

**Reply by Ariane-B** [reply]
I'd expect them to tell you your code is invalid and couldn't be compiled. 

Anyway extreme character-saving techniques like that aren't used by everyone. Those who want to use them can just disable the option, I guess.
